text/0000-prepare-phase.md

```diff
@@ -0,0 +1,105 @@
+# Meta
+[meta]: #meta
+- Name: Prepare Phase
+- Start Date: 2021-01-13
+- Author(s): [@jkutner](github.com/jkutner/)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: [buildpacks/spec#176](https://github.com/buildpacks/spec/pull/176)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal for a new Lifecycle phase, called "prepare", that would run before all other phases.
+
+# Definitions
+[definitions]: #definitions
+
+* __stack descriptor__ - an enhanced version of the [`stack.toml`](https://github.com/buildpacks/spec/blob/main/platform.md#stacktoml-toml)
+* __project descriptor__ - the [`project.toml`](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md) extension specification
+
+# Motivation
+[motivation]: #motivation
+
+The prepare phase would support the following features and capabilities:
+* [Stack buildpacks](https://github.com/buildpacks/rfcs/pull/111), which require a phase to read run-image mixins validation prior to detection
+* [Inline buildpacks](https://github.com/buildpacks/rfcs/blob/main/text/0048-inline-buildpack.md), which require parsing of the `project.toml` in the lifecycle
+* [Lifecycle configuration](https://github.com/buildpacks/rfcs/pull/128)
+
+# What it is
+[what-it-is]: #what-it-is
+
+The prepare phase will run before all other phases, and prepare the execution environment for a buildpack build. This phase will have access to secrets and credentials used to access registries and other services.
+
+## Responsibilities
+
+* Stack validation, to ensure that a new run-image is campatible with the previous app image
+* Retrive run-image mixins, which will be use dby subsequent phases
+* Validation of registry credentials, to avoid a long build that fails during export phase
+* Parsing the project descriptor and performance various operations based on its contents, include:
+    - downloading buildpacks
+    - creating ephemeral buildpacks
+    - applying include and exclude rules
+    - modifying environment variables
+
+## Inputs
+
+* Log level
+* Run-image
+* Stack ID
+* Project descriptor (optional)
+* App source code
+
+## Output
+
+* Exit status
+* Info-level logs to `stdout`
+* Error-level logs to `stderr`
+* Stack descriptor
+* Analysis metadata
+* Buildpacks (derived from inline buildpacks in project descriptor, or buildpacks in project descriptor that are not present in the builder)
+* Order definition (derived from buildpacks configuration in project descriptor)
+* Lifecycle configuration (derived from configuration in project descriptor)
+* Mutated app source code (applying include and exclude rules in project descriptor)
```

Comment:
Should we output [project metadata](https://github.com/buildpacks/spec/blob/main/platform.md#project-metadatatoml-toml)?
```suggestion
* Mutated app source code (applying include and exclude rules in project descriptor)
* Project metadata 
```
I imagine this would eventually be enhanced with other fields like `authors` and `documentaion-url` from `project.toml`. We could probably skip it for now but eventually we may want to add source metadata detection logic into the lifecycle as well.

Reply:
Bumped to future work (all of project.toml)